Change ClassHelper to use Maps instead of set

krmahadevan opened the pull request:
Fixes #1339

### Did you remember to?

- [ ] Add test case(s)
- [X] Update `CHANGES.txt`

krmahadevan commented:
ping @juherr . I am not sure how to add tests that will measure performance. So I haven't added any tests. But since this a core method that is used by TestNG, I believe that if the build turns green, that itself should be sufficient to confirm that functionality is not broken. Please help take a look.

juherr commented:
@nathanila Could you check if the change fixes your issue?

nathanila commented:
@juherr I would love to check.  How do I get the jar file?

juherr commented:
@nathanila You can build it from https://github.com/krmahadevan/testng/tree/krmahadevan-fix-1339 or maybe @krmahadevan can post it here.

cbeust commented on src/main/java/org/testng/internal/ClassHelper.java:
```diff
@@ -165,16 +166,26 @@ private static void logClassNotFoundError(String className, Exception ex) {
+        methods.get(method.getName()).add(method);
+        continue;
+      }
+      methods.put(method.getName(), new HashSet<Method>() {{add(method);}});
```
Nitpick: I'd prefer reversing the logic and not use `continue`:

    String name = method.getName();
    Set<Method> methods = methodsByNames.get(name);
    if (methods == null) {
        methods = new HashSet<>();
        methodsByNames.put(name, methods);
    }
    methods.add(method);

juherr replied:
+1

krmahadevan replied:
@cbeust / @juherr   any particular reason behind why you guys would say so ? Wouldn't want to lose an opportunity to learn :)

juherr replied:
https://blog.jooq.org/2014/12/08/dont-be-clever-the-double-curly-braces-anti-pattern/

cbeust commented on src/main/java/org/testng/internal/ClassHelper.java:
```diff
@@ -224,20 +235,28 @@ public static IJUnitTestRunner createTestRunner(TestRunner runner) {
   }
 
-  private static boolean isOverridden(Method method, Set<Method> collectedMethods) {
+  private static boolean isOverridden(Method method, Map<String, Set<Method>> methods) {
```
Agreed.

krmahadevan replied:
Fixed.

krmahadevan commented on src/main/java/org/testng/internal/ClassHelper.java:
```diff
@@ -224,20 +235,28 @@ public static IJUnitTestRunner createTestRunner(TestRunner runner) {
+            methods.get(method.getName()).add(method);
+            continue;
+          }
+          methods.put(method.getName(), new HashSet<Method>() {{add(method);}});
```
@juherr - I didn't quite understand that part. That line will be executed only once when the map initially does not contain any entries for a particular method name. So its only once per method name. 
> Avoid {{}} because it is creating many anonymous sub-HashSet instances.

Can you please help me understand this comment ?

krmahadevan commented on src/main/java/org/testng/internal/ClassHelper.java:
```diff
@@ -224,20 +235,28 @@ public static IJUnitTestRunner createTestRunner(TestRunner runner) {
 
-  private static boolean isOverridden(Method method, Set<Method> collectedMethods) {
+  private static boolean isOverridden(Method method, Map<String, Set<Method>> methods) {
+    Set<Method> collectedMethods = methods.get(method.getName());
```
Fixed.

krmahadevan commented:
> Testing for performance is not trivial but at the very least, you could add a test that makes sure the current behavior is equivalent to the previous one.

@cbeust - I didn't add an explicit method for this because I believe TestNG internally uses this method in a lot of places and the build is passing through.

krmahadevan commented:
@juherr - I have addressed the review comments.

krmahadevan commented:
@nathanila - You can get the built jar from [here](https://github.com/krmahadevan/testng/releases/tag/1339)

juherr commented on src/main/java/org/testng/internal/ClassHelper.java:
```diff
@@ -165,16 +176,25 @@ private static void logClassNotFoundError(String className, Exception ex) {
-    methods.addAll(Arrays.asList(clazz.getDeclaredMethods()));
+    Map<String, Set<Method>> methods = Maps.newHashMap();
+    for (final Method method : clazz.getDeclaredMethods()) {
+      if (! methods.containsKey(method.getName())) {
```
The pattern provided by @cbeust is a bit more optimized because you look for only once.
Here, you look for twice.

krmahadevan commented:
@juherr - I have addressed your review comments .
@cbeust - I added a test for the change as well.

Please let me know if this is now good to be merged.

krmahadevan commented:
The failure 

```
TestNG > DataProvider > test.dataprovider.DataProviderTest.parallelDataProviderSample FAILED
    java.lang.AssertionError at DataProviderTest.java:185
```

is not related to my change

krmahadevan commented:
ping @cbeust . @juherr has approved this PR. Please help merge, if you are fine with the changes.

nathanila commented:
I tried out the optimized jar.  ClassHelper.isOverridden() no longer consumes time in my profile.  Thank you for this optimization.
